app/controllers/kits_controller.rb

```diff
@@ -69,21 +63,14 @@ def allocate
     @kit = Kit.find(params[:id])
     @storage_location = current_organization.storage_locations.active_locations.find(kit_adjustment_params[:storage_location_id])
     @change_by = kit_adjustment_params[:change_by].to_i
-
-    if @change_by.positive?
-      service = AllocateKitInventoryService.new(kit: @kit, storage_location: @storage_location, increase_by: @change_by)
-      service.allocate
-      flash[:error] = service.error if service.error
-    elsif @change_by.negative?
-      service = DeallocateKitInventoryService.new(kit: @kit, storage_location: @storage_location, decrease_by: @change_by.abs)
-      service.deallocate
-      flash[:error] = service.error if service.error
-    end
-
-    if service.error
-      flash[:error] = service.error
-    else
-      flash[:notice] = "#{@kit.name} at #{@storage_location.name} quantity has changed by #{@change_by}"
+    begin
+      if @change_by.positive?
+        KitAllocateEvent.publish(@kit, @storage_location.id, @change_by)
+      else
+        KitDeallocateEvent.publish(@kit, @storage_location.id, -@change_by)
+      end
+    rescue => e
```

Comment:
Some small risk that this will catch something that it shouldn't, yes? But the user will report it if need be.